framework/aster-frame/src/mm/io.rs

```diff
@@ -168,28 +176,160 @@ impl_vmio_pointer!(&mut T, "(**self)");
 impl_vmio_pointer!(Box<T>, "(**self)");
 impl_vmio_pointer!(Arc<T>, "(**self)");
 
-/// VmReader is a reader for reading data from a contiguous range of memory.
-pub struct VmReader<'a> {
+/// A marker structure used for [`VmReader`] and [`VmWriter`],
+/// representing their operated memory scope is in user space.
+pub struct UserSpace;
+
+/// A marker structure used for [`VmReader`] and [`VmWriter`],
+/// representing their operated memory scope is in kernel space.
+pub struct KernelSpace;
+
+struct CopyInst<From, To> {
+    phantom: PhantomData<(From, To)>,
+}
+
+trait CopyMethod {
+    /// Copies `len` bytes from `src` to `dst`. Returns the number of successfully copied bytes.
+    ///
+    /// # Safety
+    ///
+    /// `src` and `dst` must meet one of the following requirements:
+    /// - Both of [`src`..`src` + len] and [`dst`..`dst` + len] are in kernel space and are
+    ///   ensured to be valid.
+    /// - One of [`src`..`src` + len] and [`dst`..`dst` + len] is in user space and the other
+    ///   memory scope is in kernel space and is ensured to be valid. In this case this method
+    ///   should be implemented with recoverable copying instructions and only be invoked when
+    ///   a suitable page table is activated.
+    ///
+    /// Otherwise, this method should not perform any instruction and return 0 directly.
+    ///
+    /// The region of memory [`src`..`src` + len] and [`dst`..`dst` + len] must _not_ overlap if
+    /// one of them represent typed memory.
+    unsafe fn copy(src: &mut *const u8, dst: &mut *mut u8, len: usize) -> usize;
+
+    /// Copies `len` bytes from `src` to `dst`.
+    /// If can not fully copy `len` bytes, returns `Err()`.
+    ///
+    /// # Safety
+    ///
+    /// `src` and `dst` must meet the safety the same requirements as [`CopyMethod::copy`].
+    unsafe fn copy_all(src: &mut *const u8, dst: &mut *mut u8, len: usize) -> Result<()>;
+}
```

Comment:
I found the generalization of the trait `CopyMethod` is awkward, as it tends to reuse code but actually reuses nothing (during specialization). Also, it mixes the concept of unsafe implementation requirement and the unsafe usage requirement. Part of the safety requirement is for the implementer and the other part is for the caller.

Here's the revised trait definition addressing the concept (writing good documentation for bad code):

```rust
/// A general implementation interface for readers and writers from various combinations of
/// virtual address spaces.
///
/// # Safety
///
/// There are different safety requirements of implementation for different spectrum of inputs:
///  - If both the source slice (`src..src + len`) and the destination slice (`dst..dst + len`) should be in kernel space,
///    The implementer should ensure validity copying overlapping ranges. And the safety requirement should
///    be the same as [`core::ptr::copy`].
///  - If exactly one of the source slice and the destination slice should be in user space and the other one
///     is in kernel space, copying should be safe. In this case methods
///     should be implemented with recoverable copying instructions and only be invoked when
///     a suitable page table is activated.
unsafe trait CopyMethod {
    /// Copies `len` bytes from `src` to `dst`. Returns the number of successfully copied bytes.
    ///
    /// # Safety
    /// 
    /// The region of memory [`src`..`src` + len] and [`dst`..`dst` + len] must _not_ overlap if
    /// one of them represent typed memory.
    ///
    /// Also, operation on typed memory may be safe only if it is plain-old-data. Otherwise you
    /// should also consider all safety requirements of [`core::ptr::copy`].
    unsafe fn copy(src: &mut *const u8, dst: &mut *mut u8, len: usize) -> usize;

    /// Copies `len` bytes from `src` to `dst`.
    /// If can not fully copy `len` bytes, returns `Err()`.
    ///
    /// # Safety
    ///
    /// `src` and `dst` must meet the safety the same requirements as [`CopyMethod::copy`].
    unsafe fn copy_all(src: &mut *const u8, dst: &mut *mut u8, len: usize) -> Result<()>;
}
```

My suggestion is: remove `CopyMethod` and `CopyInst`. Just implement 3 functions that does the same thing:
 - `copy_user_to_kernel(src, dst, len)`;
 - `copy_kernel_to_user(src, dst, len)`;
 - `copy_kernel_to_kernel(src, dst, len)`;

Also, remember to remove the incomplete trait `specialization` after this. XD

Reply:
> My suggestion is: remove CopyMethod and CopyInst. Just implement 3 functions that does the same thing:
copy_user_to_kernel(src, dst, len);
copy_kernel_to_user(src, dst, len);
copy_kernel_to_kernel(src, dst, len);

`CopyMethod` and `CopyInst` are mainly used to reuse code in `impl<'a, Target> VmReader<'a, Target>` and `impl<'a, Target> VmWriter<'a, Target>`. With them, The `read` and `write` methods can be uniformly implemented for `VmReader` and `VmWriter` without the need for multiple implementations based on different generic parameters. Furthermore, this allows `read_val` and `write_val` to directly utilize the `read` and `write` methods. If `read` and `write` were to be implemented separately for each generic parameter, `read_val` and `write_val` would also require separate implementations since the struct does not have default implementations of the `read` and `write` methods.

Reply:
I mean, you can use [TypeId](https://doc.rust-lang.org/beta/core/any/struct.TypeId.html) for this purpose. For example:

```rust
    pub fn read<W>(&mut self, writer: &mut VmWriter<'_, W>) -> usize {
        let copy_len = self.remain().min(writer.avail());
        if copy_len == 0 {
            return 0;
        }
        if TypeId::of::<W>() == TypeId::of::<UserSpace>() && TypeId::of::<Target>() == TypeId::of::<UserSpace>() {
            panic!("currently not allowed to copy from user to user")
        } else if TypeId::of::<W>() == TypeId::of::<UserSpace>() {
            unsafe { copy_kernel_to_user(&mut self.cursor, &mut writer.cursor, copy_len) }
        } else if TypeId::of::<Target>() == TypeId::of::<UserSpace>() {
            unsafe { copy_user_to_kernel(&mut self.cursor, &mut writer.cursor, copy_len) }
        } else { // both `W` and `Target` are in the kernel space.
            unsafe { copy_kernel_to_kernel(&mut self.cursor, &mut writer.cursor, copy_len) }
        }
    }
```

The compiler **IS** smart enough to optimize the branch out since `W` and `Target` are known at compile-time. You can get a confirmation here: [the compiler explorer snippet](https://godbolt.org/z/o6GrMYfxz).

Reply:
Got it!